src/simsopt/objectives/constrained.py

```diff
@@ -0,0 +1,249 @@
+# coding: utf-8
+# Copyright (c) HiddenSymmetries Development Team.
+# Distributed under the terms of the LGPL License
+
+"""
+Provides the ConstrainedProblem class implemented using the graph based
+optimization framework.
+"""
+
+from __future__ import annotations
+
+import logging
+import warnings
+from typing import Union, Callable, Tuple, Sequence
+from numbers import Real
+
+import numpy as np
+
+from .._core.optimizable import Optimizable
+from .._core.util import ObjectiveFailure
+from .._core.types import RealArray
+
+__all__ = ['ConstrainedProblem']
+
+logger = logging.getLogger(__name__)
+
+
+class ConstrainedProblem(Optimizable):
+    """
+    Represents a nonlinear, constrained optimization problem implemented using the 
+    graph based optimization framework. A ConstrainedProblem instance has
+    4 basic attributes: an objective (`f`), nonlinear constraints (`c`), 
+    linear constraints, and bound constraints. Problems take the general form:
+
+    .. math::
+
+        \min_x f(x)
+
+        s.t.
+
+          l_c \le c(x) \le u_c
+
+          Ax \le b
+
+          l_x \le x \le u_x
+
+
+    Args:
+        f_obj: objective function handle (generally one of the output functions of
+            the Optimizable instances)
+        tuples_nlc: Nonlinear constraints as a sequence of triples containing 
+            the nonlinear constraint function c with lower and upper bounds
+            i.e. `(c,l_c,u_c)`.
+            Constraint handle can (`c`) can be vector-valued or scalar-valued.
+            Constraint bounds can also be array or scalar.
+            Use ``+- np.inf`` to indicate unbounded components.
+            Define equality constraints by using equal upper and lower bounds.
+        tuple_lc: Linear constraints as a tuple containing the 2d-array A, and 1d array
```

Comment:
why do nonlinear constraints have lower & upper bounds, but linear constraints do not?

Reply:
are these linear constraints equality, or inequality? could the documentation be clarified?

Reply:
good point. I am changing the linear constraints to take lower and upper bounds, and clarifying the use of equality or inequality.